src/pcap/bosh.go

```diff
@@ -1,55 +1,414 @@
+	}
+
+	return response, res.StatusCode, nil
+}
+
+// UaaKeyInfo holds the response of the UAA /token_keys endpoint.
+type UaaKeyInfo struct {
+	Kty   string `json:"kty"`
+	E     string `json:"e"`
+	Use   string `json:"use"`
+	Kid   string `json:"kid"`
+	Alg   string `json:"alg"`
+	Value string `json:"value"`
+	N     string `json:"n"`
+}
+
+// verifyJWT checks the JWT token in tokenString and ensures that it's valid and contains the neededScope as claim.
+// Validity is determined with the defaults, i.e.
+//   - validity time range
+//   - for RSA signed JWT that the RSA signature is consistent with the key provided by UAA
+//   - that there is a claim 'scope' that contains one entry that matches neededScope.
+//
+// Limitations: only RSA signed tokens are supported.
+//
+// returns a boolean that confirms that the token is valid, from a valid issuer and has the needed scope,
+// and an error in case anything went wrong while verifying the token and its scopes.
+func (br *BoshResolver) verifyJWT(tokenString string) (bool, error) {
+	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
+		if jku, ok := token.Header["jku"]; ok {
+			jkuURL, err := url.Parse(jku.(string))
+			if err != nil {
+				return nil, err
+			}
+
+			for _, issuer := range br.uaaURLs {
+				var issuerURL *url.URL
+				issuerURL, err = url.Parse(issuer)
+				if err != nil {
+					log.Warnf("could not parse URL %s: %v", issuer, err)
+					continue
+				}
+
+				if strings.HasPrefix(jkuURL.String(), issuerURL.String()) {
+					return br.parseRsaToken(token)
+				}
+			}
+			return nil, fmt.Errorf("header 'jku' %v did not match any UAA base URLs reported by the BOSH Director: %v", jku, br.uaaURLs)
+		}
+		return nil, fmt.Errorf("header 'jku' missing from token, cannot verify signature")
+	})
+
+	if err != nil || !token.Valid {
+		return false, err
+	}
+
+	if claims, claimsOk := token.Claims.(jwt.MapClaims); claimsOk {
+		if scopes, ok := claims["scope"].([]interface{}); ok {
+			for _, scope := range scopes {
+				if scope.(string) == br.config.TokenScope {
+					return true, nil
+				}
+			}
+		}
+	}
+
+	return false, fmt.Errorf("could not find scope %q in token claims", br.config.TokenScope)
+}
+
+// parseRsaToken uses the token information for RSA signed JWT tokens and retrieves
+// the public key information from the 'jku' header in order to retrieve key information
+// (key ID, RSA public key), which is used to verify the token.
+//
+// Limitation: only supports RSA tokens using the 'jku' header, which points to a URL
+// that can be used to retrieve key information.
+func (br *BoshResolver) parseRsaToken(token *jwt.Token) (interface{}, error) {
+	if rsa, ok := token.Method.(*jwt.SigningMethodRSA); ok {
+		// with the RSA signing method, the key is a public key / certificate that can be
+		// retrieved from the JKU endpoint (among other places).
+		if rawKeyInfoURL, ok := token.Header["jku"].(string); ok {
+			var kid string
```

Comment:
The same as above. I would prefer to check if it is not ok and return in this case. And continue with non-error flow.